[FIX] mass_mailing: allow user to delete attachment after sending test email

lecr-odoo opened the pull request:
### Description of the issue this PR addresses:
When a user lacking in Administration Access Rights tries to delete an attachment after sending out a test email, it throws a traceback. This is due to the _get_with_access function trying to call the _get_mail_message_access on the res.users model. This function isn't implemented for res.users. 

This commit https://github.com/odoo/odoo/commit/8a9981e78a438ceb53a4fdc0445ebed2170c6624 added the res.users as the model into the list of values being passed in when sending out a test email to allow links to be replaced inside the emails being sent out. The primary field to allow that to happen was the res_id field. Removing the model from the list of values still allows for the links to be replaced correctly. 

This commit would fix the traceback occurring due to the _get_with_access function trying to call an unimplemented function on the res.users model and allow the user to delete the attachment after sending out a test email.

opw-4621212

### Steps to reproduce issue on Runbot:
1. Create a new user in Settings -> Users. This can be a  duplicate of Mitchell Admin
2. Change the “Administration” rights to None (i.e choose blank in the dropdown)
3. Change password to something so we can log in on an incognito browser.
4. Open an incognito browser and login as the new user.
5. Go to Email Marketing and create a new mailing. Add a sample document in the attachment in the settings tab of this mailing. Save it.
6. Test this mailing and send the test email
7. Now click on the attachment logo next to the activities button and try to delete the attachment. The traceback appears.
8. If we try to delete the attachment before sending the test email, we are able to delete it without error.
9. If we select either “Access Rights” or “Settings” in the Administration rights for this user, we do not get the error.



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/210379.png)](https://mergebot.odoo.com/odoo/odoo/pull/210379)

lecr-odoo commented:
Hi @Guillaume-gdi, could you please review this fix when you get the chance?

Guillaume-gdi reviewed (changes requested):
- Can you please make sure the lines in your commit message don't exceed 72 char.
- Can you add a test
- Great to have the commit that introduced the issue :+1: Can you link it like that:
```
[This commit] added the res.users ...
...
...
[This commit]: https://github.com/odoo/odoo/commit/8a9981e78a438ceb53a4fdc0445ebed2170c6624
```
- Can you add steps to reproduce the issue from a blank database in the commit message.

lecr-odoo commented:
@Guillaume-gdi 
> * Can you please make sure the lines in your commit message don't exceed 72 char.
Edited the commit message so that they don't exceed 72 chars, is it alright if I leave the link in or should I remove that from the commit message?
> * Can you add a test
There was already a [test](https://github.com/odoo/odoo/commit/8a9981e78a438ceb53a4fdc0445ebed2170c6624#diff-327a41968f8e6034dc784940c3f46614c2d16f296df8bcd59133ab62a7328be8R78) written for this, should I still write a test?
> * Great to have the commit that introduced the issue 👍 Can you link it like that:
> 
> ```
> [This commit] added the res.users ...
> ...
> ...
> 
> [This commit]: https://github.com/odoo/odoo/commit/8a9981e78a438ceb53a4fdc0445ebed2170c6624
> 
> opw-4621212
> ```
> 
Updated the PR comment for this.
> * Can you add steps to reproduce the issue from a blank database in the commit message.
Added this in the commit message.

lecr-odoo commented:
> You can check how we link other commit in existing commit messages like [this one](https://github.com/odoo/odoo/pull/145465/commits/a67480960085a553e97400f5ec4d740b8ea71497) for example.

Updated the commit message to look like this.

> You are fixing something that was not tested I suppose. Otherwise the test would fail before this PR.

I've added in a test to check for the deletion of the attachment on a user with the same type of permission which was causing the issue.

Guillaume-gdi reviewed (changes requested):
thanks for adding the test!

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +167,44 @@ def test_mailing_test_equals_reality(self):
             "Should use the value of the previous record's email_to as default",
         )
+
+    # Testing that an internal user without being an admin is able to delete attachments
+    # after sending a test email
```
Moved that into the docstring

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +167,44 @@ def test_mailing_test_equals_reality(self):
+    # after sending a test email
+    def test_mailing_test_correct_model(self):
+        mailing = self.test_mailing_bl.with_env(self.env)
+        # Creating the attachment and adding it mailing
```
Removed extra comments in the test

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -31,6 +32,14 @@ def setUpClass(cls):
+            login='user_not_admin',
+            password='test1234',
+            name='Martial Only Mailing',
+            signature='--\nMartial',
```
Removed that and updated the name on the user

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -5,6 +5,7 @@
 from odoo.addons.test_mass_mailing.tests.common import TestMassMailCommon
 from odoo.fields import Command
 from odoo.tests.common import users, tagged
+from odoo.addons.mail.tests.common import mail_new_test_user
```
Got it, removed the import from here

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +167,44 @@ def test_mailing_test_equals_reality(self):
+        attachment["access_token"] = attachment._generate_access_token()
+
+        # Logging in as the user and then sending the test email
+        self.authenticate(user="user_not_admin", password="test1234")
+        mailing_test = self.env['mailing.mailing.test'].create({
+            'email_to': 'test@test.com',
```
I did not know that was a thing, I've updated that and will keep that in mind for the future.

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +167,44 @@ def test_mailing_test_equals_reality(self):
+    def test_mailing_test_correct_model(self):
+        mailing = self.test_mailing_bl.with_env(self.env)
+        # Creating the attachment and adding it mailing
+        attachment = self.env['ir.attachment'].sudo().create({
```
Did not need to do that, that was leftover code, removed it.

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +167,44 @@ def test_mailing_test_equals_reality(self):
+            params={
+                "attachment_id": attachment["id"],
+                "access_token": attachment["access_token"],
+            },
```
I'm testing it through the controller function since that's the one used when deleting attachments and I wrote this test with that flow in mind. The main issue from this bug comes from the ```_get_with_access check``` which happens in the ```mail_attachment_delete``` function where the access error and then the traceback occurs.

I've added the checked to test if the attachment was deleted or not.

@Guillaume-gdi

Guillaume-gdi commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +165,39 @@ def test_mailing_test_equals_reality(self):
+                'access_token': attachment['access_token'],
+            },
+        )
+        self.assertFalse(self.env['ir.attachment'].sudo().search([('id', '=', attachment.id)]))
```
- why sudo again? :thinking: 
- why not using `browse()`?

lecr-odoo replied:
My bad, removed that.

Used search since that seemed more convenient to me at the time. Switched it to browse. One question here on using browse, does it pull data from the cache?

@Guillaume-gdi

Guillaume-gdi replied:
It returns a recordset for the ids provided as parameter in the current environment. But be aware that it doesn't check if it exists :wink: 

![image](https://github.com/user-attachments/assets/29edab89-6174-451e-afd6-c85b72d3e903)

lecr-odoo commented:
@Guillaume-gdi not sure why that error occurred, there was some [discussion](https://runbot.odoo.com/odoo/runbot.build.error/70404) about that but running the tests again worked.

lecr-odoo commented:
Hi @odoo/rd-sm-engagement, @odoo/rd-sm-phishing, could you please review this PR when you get the time? Thank you!

lecr-odoo commented on addons/mass_mailing/wizard/mailing_mailing_test.py:
```diff
@@ -80,7 +80,6 @@ def send_mail_test(self):
                 'attachment_ids': [(4, attachment.id) for attachment in mailing.attachment_ids],
                 'auto_delete': False,  # they are manually deleted after notifying the document
                 'mail_server_id': mailing.mail_server_id.id,
-                'model': 'res.users',
```
When you mention the default implementation of ```_get_mail_message_access``` do you mean this part:
```
            if hasattr(documents, '_get_mail_message_access'):
                doc_operation = documents._get_mail_message_access(docid_msgids, operation)  # why not giving model here?
            else:
                doc_operation = self.env['mail.thread']._get_mail_message_access(docid_msgids, operation, model_name=model)
            doc_result = documents._check_access(doc_operation)
```

If I were to implement something similar to this so that ```res.users``` models were taken care of, wouldn't the issue of the user needing access to the ```res.users``` model still persist? Since it doesn't make sense that a user is able to delete an attachment before sending out a test email but is unable to do so after that because the ```res.users``` model gets attached onto the record. 
I tried searching for this before as well, but I haven't seen any other examples of the ```res.users``` model being added ```mail.mail``` record when creating one.

lecr-odoo replied:
Hi @reth-odoo, my apologies for the late reply, I got caught up with some other work. I'm still looking into how to have the behavior used in _prepare_mail_values_dynamic be mimic'd in this case.
Even if I were to implement this, the model that is being attached to the attachment, which is res.users would still cause an issue here right unless I implement a _get_mail_message_access for the model?
Also would implementing the _get_mail_message_access for res.users just for this one instance make sense, since I've only seen the res.users model being passed in as a value only when sending out test emails via the mass_mailing module. Wouldn't it maybe be better to change the model being used when sending out this test email to mail.message, since over here the composer isn't being used when sending out this test email but the attachment from the chatter is being affected?

reth-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +165,39 @@ def test_mailing_test_equals_reality(self):
+                'access_token': attachment['access_token'],
+            },
+        )
+        self.assertFalse(self.env['ir.attachment'].browse(attachment.id).exists())
```
You can just do `attachment.exists()`

lecr-odoo replied:
Updated that to use attachment.exists()

lecr-odoo replied:
Hi @reth-odoo, just pinging here to check if everything is fine with the PR after the changes I've made following tde's comments?

reth-odoo reviewed (approved):
@tde-banana-odoo yet another "copy the attachments before sending them" PR

tde-banana-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +165,39 @@ def test_mailing_test_equals_reality(self):
             "Should use the value of the previous record's email_to as default",
         )
+
+    def test_mailing_test_correct_model(self):
```
:socks: Order tests alphabetically in a given class (as it is the execution order, better to follow it)

tde-banana-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -158,3 +165,39 @@ def test_mailing_test_equals_reality(self):
+        attachment['access_token'] = attachment._generate_access_token()
+
+        self.authenticate(user='user_not_admin', password='test1234')
+        mailing_test = self.env['mailing.mailing.test'].create({
```
Would it be possible to use real users ? Here everything is done using root, which is a bit sad when dealing with attachments :)

lecr-odoo replied:
Hi @tde-banana-odoo, just pinging again on this

lecr-odoo commented on addons/test_mass_mailing/tests/test_mailing_test.py:
```diff
@@ -114,6 +121,43 @@ def test_mailing_test_button_links(self):
+                'access_token': attachment['access_token'],
+            },
+        )
+        self.assertFalse(attachment.exists())
```
I've added that check above, I check that the attachment on the mail getting sent out is different from the original attachment and that the model on that is res.model, I've used sudo in the search since user_marketing doesn't have the access to the mail.mail model

lecr-odoo replied:
Hi @tde-banana-odoo  just pinging again on this.

tde-banana-odoo replied:
If I may be annoying (well, a bit more), can't we include attachment test in an existing test, so that we don't define a complete test just for that ? I suppose we could add it in 'test_mailing_test_button', after the first send_mail_test, just check attachments stuff :)
